Task fail on standard error

### Where are you running it?
Azure DevOps Services

### Expected behaviour and actual behaviour
My Pester task seems to fail because the program under test write to the standard error.
All tests passed
```
Tests completed in 59.3s


Tests Passed: 4, 
Failed: 0, 
Skipped: 0 
NotRun: 0


Pester Script finished
Finishing: Pester
```
but because my program write to the error stream (the program writes an error message if the available disk space is less than 10GiB)
```
##[error]LOW DISK SPACE.
```
the whole task fail.

I don't know if it really qualify as a bug but I'd love to have an option failOnStderr like with the PowerShell task.
In the meantime, I set `continueOnError: True`.

Thanks,
Sam

--- a/Extension/PesterTask/PesterV10/Pester.ps1
+++ b/Extension/PesterTask/PesterV10/Pester.ps1
@@ -40,13 +40,17 @@ param
 
     [string]$ScriptBlock,
 
-    [string]$TargetPesterVersion = "latest"
+    [string]$TargetPesterVersion = "latest",
+    
+    [string]$FailOnStdErr
 )
 
 if ($TargetPesterVersion -match '^4') {
     Write-Host "##vso[task.logissue type=error]This version of the task does not support Pester V4, please use task version 9."
     exit 1
 }
+
+$FailStdErr = [Boolean]::Parse($FailOnStdErr)
 
 Write-Host "TestFolder $TestFolder"
 Write-Host "resultsFile $resultsFile"
@@ -57,6 +61,7 @@ Write-Host "ExcludeTag $ExcludeTag"
 Write-Host "CodeCoverageOutputFile $CodeCoverageOutputFile"
 Write-Host "CodeCoverageFolder $CodeCoverageFolder"
 Write-Host "ScriptBlock $ScriptBlock"
+Write-Host "FailOnStdErr $FailOnStdErr"
 
 Import-Module -Name (Join-Path $PSScriptRoot "HelperModule.psm1") -Force
 Import-Pester -Version $TargetPesterVersion
@@ -142,7 +147,12 @@ if (-not([String]::IsNullOrWhiteSpace($ScriptBlock))) {
     $ScriptBlockObject.Invoke()
 }
 
-$result = Invoke-Pester -Configuration  ([PesterConfiguration]$PesterConfig)
+if ($FailStdErr) {
+    $result = Invoke-Pester -Configuration  ([PesterConfiguration]$PesterConfig)
+}
+else {
+    $result = Invoke-Pester -Configuration  ([PesterConfiguration]$PesterConfig) 2> $null
+}
 
 if ($Result.Failed.Count -gt 0) {
     Write-Error "Pester Failed at least one test. Please see results for details."

--- a/Extension/PesterTask/PesterV10/task.json
+++ b/Extension/PesterTask/PesterV10/task.json
@@ -145,6 +145,13 @@
       "required": false,
       "helpMarkDown": "A scriptblock to run before tests are executed. This will be run in the scope of the task where Invoke-Pester is run.",
       "groupName": "advanced"
+    },
+    {
+      "name": "failOnStdErr",
+      "type": "boolean",
+      "defaultValue": true,
+      "label": "Fail on StdErr",
+      "helpMarkDown": "Sets whether the task should fail on any output to StdErr from within the tests or not. Default true"
     }
    ],
   "instanceNameFormat": "Pester Test Runner",

--- a/Extension/PesterTask/PesterV9/Pester.ps1
+++ b/Extension/PesterTask/PesterV9/Pester.ps1
@@ -40,13 +40,18 @@ param
 
     [string]$ScriptBlock,
 
-    [string]$TargetPesterVersion = "latest"
+    [string]$TargetPesterVersion = "latest",
+    
+    [string]$FailOnStdErr
 )
 
 if ($TargetPesterVersion -match '^5') {
     Write-Host "##vso[task.logissue type=error]This version of the task does not support Pester V5, please use task version 10 or higher."
     exit 1
 }
+
+$FailStdErr = [Boolean]::Parse($FailOnStdErr)
+
 Write-Host "scriptFolder $scriptFolder"
 Write-Host "resultsFile $resultsFile"
 Write-Host "run32Bit $run32Bit"
@@ -56,6 +61,7 @@ Write-Host "ExcludeTag $ExcludeTag"
 Write-Host "CodeCoverageOutputFile $CodeCoverageOutputFile"
 Write-Host "CodeCoverageFolder $CodeCoverageFolder"
 Write-Host "ScriptBlock $ScriptBlock"
+Write-Host "FailOnStdErr $FailOnStdErr"
 
 Import-Module -Name (Join-Path $PSScriptRoot "HelperModule.psm1") -Force
 Import-Pester -Version $TargetPesterVersion
@@ -145,7 +151,12 @@ if (-not([String]::IsNullOrWhiteSpace($ScriptBlock))) {
     $ScriptBlockObject.Invoke()
 }
 
-$result = Invoke-Pester @Parameters
+if ($FailStdErr) {
+    $result = Invoke-Pester @Parameters
+}
+else {
+    $result = Invoke-Pester @Parameters 2> $null
+}
 
 if ($result.failedCount -ne 0) {
     Write-Error "Pester returned errors"

--- a/Extension/PesterTask/PesterV9/task.json
+++ b/Extension/PesterTask/PesterV9/task.json
@@ -142,6 +142,13 @@
       "required": false,
       "helpMarkDown": "A scriptblock to run before tests are executed. This will be run in the scope of the task where Invoke-Pester is run.",
       "groupName": "advanced"
+    },
+    {
+      "name": "failOnStdErr",
+      "type": "boolean",
+      "defaultValue": true,
+      "label": "Fail on StdErr",
+      "helpMarkDown": "Sets whether the task should fail on any output to StdErr from within the tests or not. Default true"
     }
    ],
   "instanceNameFormat": "Pester Test Runner",

--- a/Extension/PesterTask/src/pester.ts
+++ b/Extension/PesterTask/src/pester.ts
@@ -20,6 +20,7 @@ export async function run() {
         let CodeCoverageFolder = tl.getInput("CodeCoverageFolder");
         let ScriptBlock = tl.getInput("ScriptBlock");
         let PesterVersion = tl.getInput("PesterVersion");
+        let FailOnStdErr = tl.getInput("failOnStdErr");
 
         let TargetPesterVersion = "0.0.0";
         if (PesterVersion === "OtherVersion") {
@@ -48,6 +49,7 @@ export async function run() {
                     "-scriptFolder", scriptFolder,
                     "-resultsFile", resultsFile,
                     "-run32Bit", run32Bit,
+                    "-FailOnStdErr", FailOnStdErr,
                 ];
 
         if (additionalModulePath) {

--- a/Extension/PesterTask/src/pesterv10.ts
+++ b/Extension/PesterTask/src/pesterv10.ts
@@ -20,6 +20,7 @@ export async function run() {
         let CodeCoverageFolder = tl.getInput("CodeCoverageFolder");
         let ScriptBlock = tl.getInput("ScriptBlock");
         let PesterVersion = tl.getInput("PesterVersion");
+        let FailOnStdErr = tl.getInput("failOnStdErr");
 
         let TargetPesterVersion = "0.0.0";
         if (PesterVersion === "OtherVersion") {
@@ -48,6 +49,7 @@ export async function run() {
                     "-TestFolder", TestFolder,
                     "-resultsFile", resultsFile,
                     "-run32Bit", run32Bit,
+                    "-FailOnStdErr", FailOnStdErr,
                 ];
 
         if (additionalModulePath) {

